corehq/apps/locations/models.py

```diff
@@ -834,6 +837,61 @@ def from_locations(cls, locations):
         related_locations = {loc_id for relation in relations for loc_id in relation}
         return related_locations - {l.location_id for l in locations}
 
+    @classmethod
+    def relation_distance_dictionary(cls, locations):
+        """Returns a dictionary of the related locations and their distance to each
+        location that's passed in if it has a relation.
+
+        If two locations that are passed in that have a relation. That relation is ignored.
+
+        {
+          related_location: {
+            loc_a_passed_in: distance,
+            loc_b_passed_in: distance
+          }
+        }
+        """
+        relations = cls.objects.filter(
+            Q(location_a__in=locations) | Q(location_b__in=locations)
+        ).prefetch_related('location_a', 'location_b')
+        location_ids = {loc.location_id for loc in locations}
+
+        distance_dictionary = defaultdict(dict)
+
+        for relation in relations:
+            if relation.distance is None:
+                continue
+
+            a_id = relation.location_a_id
+            b_id = relation.location_b_id
+
+            if a_id in location_ids and b_id in location_ids:
+                # ignore when two locations were passed in that are related to each other
+                continue
+
+            if a_id in location_ids:
+                distance_dictionary[b_id][a_id] = relation.distance
+            elif b_id in location_ids:
+                distance_dictionary[a_id][b_id] = relation.distance
+
+        return distance_dictionary
+
+    @classmethod
+    def update_location_distances(cls, source_location_id, location_dict):
+        """Update the related locations for a source location.
+
+        location_dict: {loc_id: distance}
+        """
+
+        for loc_id, distance in six.iteritems(location_dict):
+            relation = cls.objects.filter(
+                (Q(location_a_id=loc_id) & Q(location_b_id=source_location_id)) |
+                (Q(location_b_id=loc_id) & Q(location_a_id=source_location_id))
+            ).first()
```

Comment:
Should this always return one result?  (Never 0 or more than 1)

Reply:
Yes. because it's intended to be symmetric there should only be one relation that this query could return and because we're passing in a distance dictionary, it's expected that there's at least one relation